Use c10::variant-based enums for Nonlinearity and FanMode

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27942 Use c10::variant-based enums for Reduction
* **#27933 Use c10::variant-based enums for Nonlinearity and FanMode**

Differential Revision: [D18009044](https://our.internmc.facebook.com/intern/diff/D18009044)

Commit:
Use c10::variant-based enums for Nonlinearity and FanMode

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

ezyang commented on torch/csrc/api/src/nn/init.cpp:
```diff
@@ -53,12 +105,16 @@ double calculate_kaiming_std(
-  } else if (nonlinearity == torch::nn::init::Nonlinearity::LeakyReLU) {
+  } else if (c10::get_if<enumtype::kLeakyReLU>(&nonlinearity)) {
     return std::sqrt(2.0 / (1 + pow(param, 2)));  // NOLINT
   }
```
Hmm. If we ever add a new variant to the enum, will we then get an error saying we didn't handle all cases yet? If not, we should add a fallthrough case here...

yf225 replied:
Yeah we should have a fallthrough case here - I will do it as part of the work for `torch::nn::init::calculate_gain` parity.

ezyang commented on torch/csrc/api/include/torch/nn/init.h:
```diff
@@ -21,16 +24,42 @@ enum class Nonlinearity {
   LeakyReLU
 };
 
+// This enum class is deprecated and will be removed in 1.5.
```
Hmm, I wonder if we can *statically* warn in these cases... does C10_DEPRECATED work here?

yf225 replied:
It seems that adding `C10_DEPRECATED` prevents the enum class from being used in some CI configs such as `pytorch_macos_10_13_py3_test`, hence breaking BC-support. For now I'll keep the comment instead.

ezyang commented on torch/csrc/api/src/nn/init.cpp:
```diff
@@ -35,16 +35,68 @@ struct Fan {
 
+#define COMPUTE_NONLINEARITY_ENUM(name) /* NOLINT(cppcoreguidelines-macro-usage) */ \
+case Nonlinearity::name: \
+  TORCH_WARN( \
```
`TORCH_WARN_ONCE`?

yf225 replied:
I do want people to move to the new enum style sooner than later, because the old style will be deprecated soon :D I think it's probably okay for the warning to be a bit more annoying here, since moving to the new style is minimal effort.

ezyang commented on torch/csrc/api/src/nn/init.cpp:
```diff
@@ -35,16 +35,68 @@ struct Fan {
+      TORCH_INTERNAL_ASSERT(
+        false,
+        "The enum class `torch::nn::init::Nonlinearity` is deprecated. ",
+        "Please don't add any new enum value to it.")
```
What should I do instead? ;)

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head

Commit:
Update on "Use c10::variant-based enums for Nonlinearity and FanMode"

Use c10::variant-based enums for Nonlinearity and FanMode

gh-metadata: pytorch pytorch 27933 gh/yf225/10/head
